### ISSUE TITLE
makebw generates images with alpha channel (Linux)

### ISSUE THREAD

**Issue by daniel-j** [RobtAll/novel#9]
Adding `-alpha off` to the convert command resolves the issue. Windows batch script is untested. Current script seems to work on macOS.

Would be nice to have makegray to generate a lossless png like makebw. I fixed this in my fork (and the alpha issue), see the following commit: https://github.com/daniel-j/novel/commit/0cdd96a0ae8f91fff259348e030537810046ab69

**Comment by daniel-j** [comment]
> Since you have MacOSX: Can you check what happens when a PDF input is fed to the scripts? Especially check that the resulting image is at 300dpi (makegray, makecmyk) or 800dpi (makebw). Also check that the output of makecmyk, which is a PDF file, can be opened in your Mac PDF reader. Colors might be a bit off, but that is normal (see docs).

The result looks the same as on Linux. Created pdfs opens fine. You can download the files I generated from borderland-cover.pdf here: [borderland-cover-pdf-processed-macos.zip](https://github.com/RobtAll/novel/files/1864609/borderland-cover-pdf-processed-macos.zip)

I installed imagemagick and ghostscript with [Homebrew](https://brew.sh/):
`brew install imagemagick --with-little-cms --with-little-cms2`
`brew install ghostscript`

I noticed the makegray and makebw first converts to TIF format, is that required? Why not convert to the final format directly?

**Comment by RobtAll** [comment]
Excellent. Will deal with this on Tuesday, being away from my main machine.

Why intermediate TIF? Required for makecmyk, not the others. But I did makecmyk first, and perhaps retained an unnecessary step. Or maybe it's necessary, but I forgot why!

**Comment by daniel-j** [comment]
After some testing of `makegray` I can confirm that it only needs one convert command with all options included. The final image is identical. See my changes here: https://github.com/daniel-j/novel/commit/8f237fa0c7591d209f9d849b03b61604d59dd309

**Comment by RobtAll** [comment]
Code change looks good. Will verify Monday, upload Tuesday. 

Avoid changes to makecmyk. The intermediate steps help to avoid potential colorspace issues. Besides, the added processing time is unimportant, since a book has only one cover.

Thought: should I change makegray from default jpg (option png) to default png (option jpg) as output?

**Comment by RobtAll** [comment]
Modifications made. Can now be found as novel-scripts-EXP5.zip here. Actually, I am sufficiently confident that I have already included the new scripts in novel version 1.50.2, uploaded to CTAN just now.

I mentioned that you checked Mac usage, and provided the way to get files via brew.

I also set the default makegray output to png. If the user needs jpg, there is a way to do that, documented.

In the final LuaLaTeX PDF: (1) If a placed image is jpg, then it is still jpg in the PDF. (2) If the placed image is png, it is uncompressed in the PDF. (3) Color cover CMYK is FlateDecode in the PDF. This is a lossless compression method most similar to tiff.

It seems that ImageMagick has some slight version (6 and 7) syntax changes. To be sure that the alpha channel is really removed (all scripts) I also set the background to White first, lest the alpha go to black instead.

There were also some minor internal changes, peculiar to Windows.

Will close this thread in a couple of days, unless something new emerges on the same topic.

**Comment by daniel-j** [comment]
Thanks! Here are some of my edits in my fork of novel:

makebw:
```patch
- # This is a Linux BASH command script. May also work with OSX.
+ # This is a Linux/macOS BASH command script.
...
+ # Check structure of folders and required files:
+ printf "Checking for presence of required files... "
+ if [ ! -d "temp" ]; then mkdir temp; fi
+ if [ ! -d "output" ]; then mkdir output; fi
+ 
# Checks for input arguments:
...
if [ -f "temp/temp-identify.txt" ]; then rm temp/temp-identify.txt; fi
+ if [ -d "temp" ]; then rmdir temp; fi
```

makecmyk:
```patch
- # This is a Linux BASH command script. Might work with OSX (not tested).
+ # This is a Linux/macOS BASH command script.
...
if [ -f "temp/temp-embedded.icc" ]; then rm temp/temp-embedded.icc; fi
+ if [ -d "temp" ]; then rmdir temp; fi
```

makegray:
```patch
- # This is a Linux BASH command script. May also work with OSX.
+ # This is a Linux/macOS BASH command script.
...
-   echo "Converts Color or Grayscale to single-channel Grayscale jpg."
+   echo "Converts Color or Grayscale to single-channel Grayscale png or jpg."
...
-   echo "Output will appear in output folder, named filename-P-GRAY.jpg."
+   echo "Output will appear in output folder, named filename-P-GRAY.png."
...
- echo "Converts color or grayscale image to single-channel grayscale jpg."
+ echo "Converts color or grayscale image to single-channel grayscale png or jpg."
...
+ # Check structure of folders and required files:
+ printf "Checking for presence of required files... "
+ if [ ! -d "temp" ]; then mkdir temp; fi
+ if [ ! -d "output" ]; then mkdir output; fi
+ 
# Checks for input arguments:
...
if [ -f "temp/temp-identify.txt" ]; then rm temp/temp-identify.txt; fi
+ if [ -d "temp" ]; then rmdir temp; fi
```

novel-scripts-README.html:
Add `macOS` where Linux is mentioned.

resource/portable/bin:
Is `delegates.xml` supposed to be included?

output and temp dirs should not be included, they should be generated and temp should get removed when empty (see my above commands).

**Comment by daniel-j** [comment]
_offtopic:_
I am a bit sad that the source code of novel is not available in this github repository. I made my own fork of novel a while ago with a more git-friendly structure (and some modifications/additions) that I keep up to date. Have a look if you haven't already: https://github.com/daniel-j/novel

**Comment by RobtAll** [comment]
Thanks. I incorporated some of your edits into my code and docs. Good catch creating the output and temp folders. Although they are supposed to be included (and empty) as-distributed, some folks like to discard empty folders. So all the scripts need to test and create, if necessary.

In that regard, I won't be discarding any empty temp folder. I sometimes use those things for other purposes, or even leave something in the folder if the user "might" want to see it. Currently, nothing like that is in the scripts; but it might have been that way.

The scripts will now only process the first page of an input PDF file. I added scripts to split a multi-page PDF, using Ghostscript. This is good enough, I think. Along the way, I fixed a potential bug that appears when the scripts try to gain info from a multi-page PDF: Each page reports, causing massive confusion. Now the scripts only ask the first page for info, which is all that is needed.

Unless some show-stopper emerges, the version known as EXP7 (currently in my experimental folder here) will be the version of novel-scripts.zip that is uploaded to CTAN. I have not done any work on any other parts of novel for the last few days, and don't expect it will be necessary. Upload should be Wednesday evening or Thursday morning, my time.

As for why I don't use GitHub for source code, the way that real programmers do: I am not a real programmer. I few months ago, I was in the bad habit of uploading experimental code that was wildly experimental. Anyone grabbing it from GitHub was in for a wild ride. Then I decided not to do that. If you look at my code changes over a period of time, you will see that I like to make massive revisions rather than line-item changes.

**Comment by RobtAll** [comment]
The scripts are now denoted scriptver1.2, and will be included in novel 1.50.3.

There were some changes, particularly regarding Windows. Also, a new message appears if the user attempts to process a blank pdf input. For future use, makebw and makegray add short comments; this requires an extra processing step.

**Comment by RobtAll** [comment]
Have a look in the experimental folder, at scripts version 1.4pre1.

As it turns out, it is not possible to eliminate some of the image processing steps, by condensing operations. This is due to syntax changes in various version of ImageMagick between versions 6 and 7. Depending on platform, the user may have any of several versions.

By maintaining several processing steps, it is possible to write the code so that the same end result is obtained with different IM versions (and either Windows or Linux). So, any edits to the IM commands may fail cross-platform. Be careful about changes.

I also made some changes that might be useful, if the user's print service is extremely particular about image formats. Most users will not notice a difference, and most printers won't care.

The scripts write a comment into the resulting image file. In novel version 1.52 (soon), LuaTeX will read the comment, and know whether or not the image was pre-processed by script. This is already done for color cover art. With 1.52, it will also be done for interior images. A Warning will be generated, with a list of non-scripted interior images used in the document. Should actually be an Error, but I don't want to block processing.

PNG can contain an alpha channel, and the scripts remove it if seen. But without an alpha channel, PNG can also specify a single color as fully transparent, just like GIF. I didn't know that until recently. I must investigate as to whether such transparency is removed; if not, how to remove it. There is a PNG chunk tRNS for that purpose, which I can remove if it is present, but I do not yet know the consequences of that action.
